packages/module/src/ColumnManagementModal/ColumnManagementModal.tsx

```diff
@@ -0,0 +1,146 @@
+import React from 'react';
+import {
+  Modal,
+  Button,
+  TextContent,
+  Text,
+  TextVariants,
+  DataListItem,
+  DataList,
+  DataListItemRow,
+  DataListCheck,
+  DataListCell,
+  DataListItemCells,
+  Split,
+  SplitItem,
+  ModalProps
+} from '@patternfly/react-core';
+
+export interface ColumnManagementModalColumn {
+  /** Internal identifier of a column by which table displayed columns are filtered. */
+  key: string,
+  /** The actual display name of the column possibly with a tooltip or icon. */
+  title: React.ReactNode,
+  /** If user changes checkboxes, the component will send back column array with this property altered. */
+  isShown?: boolean,
+  /** Set to false if the column should be hidden initially */
+  isShownByDefault: boolean,
+  /** The checkbox will be checked and disabled, this is applicable to identifier columns which user shouldn't be able to hide */
+  isAlwaysShown?: boolean
+}
+
+export interface ColumnManagementModalProps extends Omit<ModalProps, 'ref' | 'children'> {
+  /** Flag to show the modal */
+  isModalOpen: boolean,
+  /** Invoked when modal visibility is changed */
+  setModalOpen: (shouldOpen: boolean) => void
+  /** Current column state */
+  appliedColumns: ColumnManagementModalColumn[],
+  /** Invoked with new column state after save button is clicked */
+  applyColumns: (newColumns: ColumnManagementModalColumn[]) => void,
+};
+
+const ColumnManagementModal: React.FunctionComponent<ColumnManagementModalProps> = (
+  { isModalOpen,
+    setModalOpen,
+    appliedColumns,
+    applyColumns,
+    ...props }: ColumnManagementModalProps) => {
+
+  const [ currentColumns, setCurrentColumns ] = React.useState(
+    appliedColumns.map(column => ({ ...column, isShown: column.isShown ?? column.isShownByDefault }))
+  );
+
+  const handleChange = index => {
+    const newColumns = [ ...currentColumns ];
+    const changedColumn = { ...newColumns[index] };
+
+    changedColumn.isShown = !changedColumn.isShown;
+    newColumns[index] = changedColumn;
+
+    setCurrentColumns(newColumns);
+  };
+
+  const selectAll = () => {
+    let newColumns = [ ...currentColumns ];
+    newColumns = newColumns.map(column => ({ ...column, isShown: true }));
+
+    setCurrentColumns(newColumns);
+  };
+
+  const resetToDefault = () => {
+    setCurrentColumns(currentColumns.map(column => ({ ...column, isShown: column.isShownByDefault ?? false })));
+  };
+
+  const handleSave = () => {
+    setModalOpen(false);
+    applyColumns(currentColumns);
+  };
+
+  const handleCancel = () => {
+    setModalOpen(false);
+    setCurrentColumns(appliedColumns.map(column => ({ ...column, isShown: column.isShown ?? column.isShownByDefault })));
+  };
+
+  return (
+    <Modal
+      title="Manage columns"
```

Comment:
With DataView, this component might be used also for other layouts than table, we could make this title a prop with a default value. What do you think?